tm-monitor/docs/Design-proposal.md

```diff
@@ -0,0 +1,53 @@
+# Design proposal for Tendermint Monitor
+
+Tendermint monitor is designed to be the tool for all related parties to have a close look about the network. 
+
+## Design Goal for monitoring the network
+
+The design goal of Tendermint Monitor is to track **Liveness** & **Safety** of the whole network and certain validator.
+
+ The liveness of the network is in jeopardy if:
+
+* many validators are offline
+* the system barely gather +2/3 votes in recent blocks
+* validators have to wait for many rounds before commiting new block
+
+
+The safety of the network is in jeopardy if:
+
+* There exists a monopoly of Atom token, +1/3
+
+* Not enough Atoms are bonded
```

Comment:
please remember that it is "Tendermint monitor". Cosmos != Tendermint

Reply:
> Shall we separately propose a gaia-monitor too or shall we rebrand this for Cosmos?

no rebranding please. copy/fork is ok

Reply:
Ok, so tm-monitor will focus on Tendermint side. The validators can develop their own Cosmos monitoring tools.

Reply:
Just curious why the current monitor design is to have the clients in the monitor connect to nodes vs having the nodes report to the monitoring tool?  The latter approach is the one used by ethstats and seems to work well.   With this approach you could add a flag to `tendermint node` to direct any node to reports stats to the monitoring UI.  Ex:  `tendermint node --stats=localhost:8080 ...`  enables stats reporting and sends data to localhost:8080 (monitor). 

Here's a *rough* sketch of a service the could be included in Tendermint to allow any node to report directly to the monitor:

```
import (
	"context"
	"fmt"
	"log"
	"net/url"
	"time"

	"github.com/gorilla/websocket"
	tmtypes "github.com/tendermint/tendermint/types"
)

const (
	pingInterval = 20 * time.Second
	pongWait     = 60 * time.Second
)

type Reporter struct {
	hostport   string
	eventBus   *tmtypes.EventBus
	connection *websocket.Conn
	blockChan  chan interface{}
	quitCh     chan struct{}
}

func NewReporter(hp string, events *tmtypes.EventBus) *Reporter {
	return &Reporter{
		hostport:  hp,
		eventBus:  events,
		blockChan: make(chan interface{}),
		quitCh:    make(chan struct{}),
	}
}

func (reporter *Reporter) Start() {
	go reporter.processLoop()
}

func (reporter *Reporter) Stop() {
	reporter.eventBus.Unsubscribe(context.Background(), "TMSTATS", tmtypes.EventQueryNewBlock)
	close(reporter.quitCh)
}

func (reporter *Reporter) processLoop() {
	u := url.URL{Scheme: "ws", Host: reporter.hostport, Path: "/events"}
	log.Printf("attempting to connect to %s", u.String())

	for {
		conn, _, err := websocket.DefaultDialer.Dial(u.String(), nil)
		if err != nil {
			log.Println("Stats server unreachable try again in 10 secs...")
			time.Sleep(10 * time.Second)
			continue
		}

		// Got the connection... start the reader
		go reporter.reader(conn)

		// Subscribe to block events (maybe change to NewHeader)
		err = reporter.eventBus.Subscribe(context.Background(), "TMSTATS", tmtypes.EventQueryNewBlock, reporter.blockChan)
		if err != nil {
			conn.Close()
			return
		}

		// report loop
		pingTicker := time.NewTicker(pingInterval) // Must be < pong wait
		defer func() {
			pingTicker.Stop()
			conn.Close()
		}()

		for err == nil {
			select {
			case <-reporter.quitCh:
				conn.Close()
				return

			case <-pingTicker.C:
				//conn.SetWriteDeadline(time.Now().Add(writeWait))
				if err = conn.WriteMessage(websocket.PingMessage, []byte{}); err != nil {
					log.Println("ping error:", err)
				}
			case event := <-reporter.blockChan:
				block := event.(tmtypes.EventDataNewBlock)
				msg := fmt.Sprintf("Block: %d", block.Block.Header.Height)
				err = conn.WriteMessage(websocket.TextMessage, []byte(msg))
				if err != nil {
					log.Println("write error:", err)
				}
			}
		}
		reporter.eventBus.Unsubscribe(context.Background(), "TMSTATS", tmtypes.EventQueryNewBlock)
		conn.Close()
	}

}

func (reporter *Reporter) reader(conn *websocket.Conn) {
	defer conn.Close()

	conn.SetReadDeadline(time.Now().Add(pongWait))
	conn.SetPongHandler(func(string) error { conn.SetReadDeadline(time.Now().Add(pongWait)); return nil })
	for {
		_, msg, err := conn.ReadMessage()
		if err != nil {
			log.Println("read:", err)
			return
		}
		log.Printf("recv: %s", msg)
	}
}
```
If there's any interest in this, I also have a stub of the monitor websocket server to collect data from the nodes.  But it's written in Erlang :grimacing:

Reply:
https://prometheus.io/docs/introduction/faq/#why-do-you-pull-rather-than-push

I do not have a strong preference about pull vs push model. But with pull model, I can connect to Tendermint from 2 different tools (say Prometheus server and telegraf). While in push model, I only send metrics to one place.

Reply:
Note there is ongoing work on adding metrics to Tendermint https://github.com/tendermint/tendermint/pull/1737

Reply:
@melekes I figured there was a good reason for the approach.  Managing dozens of 'pull' clients could be a bit of a pain though: reconnect on error, etc...  

 I need to dig into the metrics stuff to learn how to access it.  Thanks.